core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -50,17 +65,68 @@ public FilteredQueryBuilder(@Nullable QueryBuilder queryBuilder, @Nullable Query
         this.filterBuilder = filterBuilder;
     }
 
+    /** Returns the query to apply the filter to. */
+    public QueryBuilder query() {
+        return queryBuilder;
+    }
+
+    /** Returns the filter to apply to the query results. */
+    public QueryBuilder filter() {
+        return filterBuilder;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(NAME);
-        if (queryBuilder != null) {
-            builder.field("query");
-            queryBuilder.toXContent(builder, params);
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder);
+    }
+
+    @Override
+    public int doHashCode() {
+        return Objects.hash(queryBuilder, filterBuilder);
+    }
+
+    @Override
+    public Query doToQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query result = null;
+
+        Query query = queryBuilder.toQuery(parseContext);
+        Query filter = filterBuilder.toQuery(parseContext);
+
+        if (query == null) {
+            return null;
```

Comment:
thinking out loud here, this is tricky: previously we could have null in case the query is there but parsed as null, which we now have EmptyQueryBuilder for. EmptyQueryBuider#toQuery returns null, so this if seems to be right here. that said I find it confusing, cause we end up ignoring query and filter in this case. Why?